src/main/java/io/jenkins/plugins/sample/OnboardingTaskBuilder.java

```diff
@@ -0,0 +1,257 @@
+                return;
+            }
+
+            Map<String, List<JobRunEntry>> categoryJobsMap;
+            try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+                Type type = new TypeToken<Map<String,List<JobRunEntry>>>(){}.getType();
+                categoryJobsMap = gson.fromJson(reader, type);
+                if (categoryJobsMap == null) {
+                    categoryJobsMap = new HashMap<>();
+                }
+            } catch (IOException | JsonSyntaxException e) {
+                categoryJobsMap = new HashMap<>();
+            }
+
+            categoryJobsMap.computeIfAbsent(category, k -> new ArrayList<>())
+                    .add(new JobRunEntry(jobName, Instant.now()));
+
+            try (BufferedWriter writer = new BufferedWriter(new FileWriter(file))) {
+                gson.toJson(categoryJobsMap, writer);
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+
+    }
+
+public static class JobRunEntry {
+    private String jobName;
+    private final Instant timestamp;
+
+    public JobRunEntry(String jobName, Instant timestamp) {
+        this.jobName = jobName;
+        this.timestamp = timestamp;
+    }
+
+    public String getJobName() {
+        return jobName;
+    }
+
+    public void setJobName(String jobName) {
+        this.jobName = jobName;
+    }
+
+    public Instant getTimestamp() {
+        return timestamp;
+    }
+}
+
+    //TODO: This approach could be slow if there too many jobs. Another way could be to maintain the last job run
+    // separate in the same json file.
+    public static String getLastJobRun() {
+        synchronized (lock) {
+            try (BufferedReader reader = new BufferedReader(new FileReader(CATEGORY_JOBS_FILE))) {
+                Type type = new TypeToken<Map<String,List<JobRunEntry>>>(){}.getType();
+                Map<String, List<JobRunEntry>> categoryJobsMap = gson.fromJson(reader, type);
+
+                if (categoryJobsMap == null || categoryJobsMap.isEmpty())
+                    return "";
+
+                JobRunEntry lastJobRun = null;
+                for (List<JobRunEntry> jobRunEntries: categoryJobsMap.values()) {
+                    for(JobRunEntry jre: jobRunEntries) {
+                        if (lastJobRun == null || lastJobRun.timestamp.isBefore(jre.timestamp)) {
+                            lastJobRun = jre;
+                        }
+                    }
+                }
+                return lastJobRun == null ? "" : lastJobRun.getJobName();
+            } catch (IOException | JsonSyntaxException e) {
+                e.printStackTrace();
+                return "";
+            }
+        }
+    }
+
+
+    public static boolean renameJobName(String oldName, String newName) {
+        synchronized (lock) {
+            File file = new File(CATEGORY_JOBS_FILE);
+            try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
```

Comment:
FWIW having all of these static methods here reading the file from disk seems a little unusual. A more typical approach might be to create a `@Extension public class OnboardingTaskJobRuns` class, have the `Map<String, List<JobRunEntry>> categoryJobsMap` be a field in that class, have instance methods in it to update/retrieve values, and have it handle saving the file internally. You would interact with it from the build by retrieving the singleton instance using `ExtensionList.lookupSingleton(OnboardingTaskJobRuns.class)`.

(Of course maybe the onboarding doc told you to use the current approach, I am not familiar with it.)

Reply:
I'll revisit this part.